### ISSUE TITLE
ENOSPC: System limit for number of file watchers reached

### ISSUE THREAD

**Issue by nicolasrouanne** [expo/expo-github-action#20]
### Description of the bug
Basically a duplicate of https://github.com/expo/expo-cli/issues/277

### To Reproduce
Run `expo publish` using default configuration in `README`, with slight modifications:
- use `yarn` instead of NPM
- use a `release-channel` option

> I don't believe these options may result in not being able to _publish_ though.

```yaml
name: Expo Review App
on:
  pull_request:
    types: [opened, synchronize, reopened]
jobs:
  publish:
    name: Install and publish
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v1
      - uses: actions/setup-node@v1
        with:
          node-version: 12.x
      - uses: expo/expo-github-action@v4
        with:
          expo-version: 3.x
          expo-username: ${{ secrets.EXPO_CLI_USERNAME }}
          expo-password: ${{ secrets.EXPO_CLI_PASSWORD }}
      - run: yarn install
      - run: expo publish --release-channel=pr-${{ github.event.number }}
```

#### Expected behavior
`expo publish` to succeed

#### Actual behavior
Following error in the `expo publish ...` step
```
ENOSPC: System limit for number of file watchers reached
```

### Additional context
I'm curious, because you state that running `expo publish` is **not working** [in the related issue](https://github.com/expo/expo-cli/issues/277#issuecomment-477057752), and that's actually what I see too.

However you also seem to succeed publishing using this action and default configuration 🤷‍♂

**Comment by nicolasrouanne** [comment]
I also tried using the [_Docker image_ configuration](https://github.com/expo/expo-github-action#use-docker-for-improved-performance) and I get the same error.

**Comment by RemusRD** [comment]
I'm running into the same problem since some weeks ago

**Comment by byCedric** [comment]
Thanks for the report guys, it's detailed and clearly written. ❤️ 

@nicolasrouanne I think I might have removed the fix from the previous report, because the images are now a bit more provider-agnostic. I'll re-add the fix in the images based on your information there. 👍 

For the GitHub Action, I'll do a bit more research how we can implement the fix. I'm not sure yet, but rest assured I'm working on it. Will make some time available this weekend to have a fix out before monday.

**Comment by nicolasrouanne** [comment]
@byCedric Thanks for the feedback and looking into it so quickly!

> ❤️ Also love the simple integration of the release channels, haven't thought of it myself yet. Do you mind if we add this as example configuration? 😄

Sure! Please do. 
If you want I can make a PR myself once I polished the angles (e.g. probably can simplify the triggers on `pull_request` event, yet to be tested).

BTW I’d also love to be able to display in the _Status check_ a link 🔗 to the expo published _Review app_ pretty much like **Netlify** is doing for their _Deploy previews_. Once I figure that out, could add it to.

**Comment by byCedric** [comment]
@nicolasrouanne That sounds awesome! If you have time go ahead, I'd be happy to review the PR 😄 

I added a small fix for the docker image inotify limits. But I can't verify if this works, because I'm not able to trigger the issue. Could you try if you still have this issue using the `bycedric/expo-cli-fix-inotify:latest` image? I might be able to "read" the value from GitHub Action, but I want to make sure this works before releasing it.

If we fixed the docker image, I'll take a look at how to "patch" GitHub Action without docker image.

**Comment by RemusRD** [comment]
@byCedric Maybe I am running in a different problem but, seems that it does not fix it 

https://github.com/RemusRD/dance-academy/runs/249184584

**Comment by byCedric** [comment]
@RemusRD Thanks for testing! ❤️ I'm afraid this seems to be a GitHub Action limitation. That docker image is forcing the `fs.inotify.max_user_watches` limit to `524288`, but apparently GHA is overwriting this back to `8192`. You can see this happen in a [fork of your repo](https://github.com/byCedric/dance-academy/runs/249245759) (when we are done, I'll remove the fork ofc, let me know if you want to have it removed earlier).

I'll create a ticket in the community, but let's keep this issue open. If I have more info in this matter, I'll update it here. My primary goal is to first fix the docker-based publish, if we have that working smoothly we should be able to fix the action/js-based publish.

**Comment by RemusRD** [comment]
@byCedric feel free to keep it open as long as it is needed! Thank you a lot for the effort! Is not this related with this issue? https://github.com/expo/expo-cli/issues/277#issuecomment-477057752

**Comment by byCedric** [comment]
Yes, it's related to a limitation of the environment you are running Expo CLI in. The metro bundler requires a high amount of listeners apparently. This fails if the host environment is limiting this. So technically its an environment issue, but I'm not sure if the CLI can change anything about this.

I find the limit in GitHub Action personally a little low. Like I tried to outline in an earlier comment on that CLI issue, the limitation in other CI vendors is actually set to the default max listeners. Why they did not do this in GH Actions is unclear, that's what I try to find out. Might be a configurational issue on their hands, or an intentional limitation.

**Comment by nicolasrouanne** [comment]
I have the [same test results](https://github.com/expo/expo-github-action/issues/20#issuecomment-538680679) as @RemusRD (thanks for testing 👌 ), still broken with `inotify`.

@byCedric I would be interested if you could link the issue you're creating here (not sure if it's an `expo-cli` of _Github Action_ issue you're creating 😉 )

**Comment by byCedric** [comment]
As the community forum of GitHub doesn't allow my message to be posted (internal server error), I'm mostly resolving this via mail. But I'll post updates here if I know more 😄 

Again, it's not necessarily an Expo CLI issue. It's a limitation in GH action that limits the number of watches on the systems. Unfortunately, there is no easy fix for this. Right now I can think of 3:

1. GitHub opens up this setting for developers to change (hard, probably need privileged mode or root-access).
2. GitHub sets the default node listeners to something higher (easy for us, hard for them probably)
3. Metro/Expo disables or limits file watchers when publishing (hard, I don't think Metro allows this kind of feature but I might be wrong there. I wish it was a feature)

**Comment by vongohren** [comment]
So, there exists a fix, that seemed to work for me when I tried. What I did was to follow this guys tip: “Increasing the number of watchers” — @JNYBGR https://link.medium.com/9Zbt3B4pM0

I then did this in my main action.yml
With all the specifics underneath the dev release
```
    steps:
    - uses: actions/checkout@v1
    - name: Setup kernel for react native, increase watchers
      run: echo fs.inotify.max_user_watches=524288 | sudo tee -a /etc/sysctl.conf && sudo sysctl -p
    - name: Run dev release fastlane inside docker action
```

I can confirm that I got past the watch step and was able to finish my dockerized fastlane build and release action

**Comment by byCedric** [comment]
@vongohren, thanks! That's interesting. That means we actually _have_ privileged access to GitHub actions? In that case I'll add that to the JS action! ♥️ For the docker version, I'm still waiting on GitHub support's response.

**Comment by vongohren** [comment]
@byCedric yeah I was not able to run this command inside docker, but I was able to run this outside, so that the docker, which you said inherits the kernel, uses the right setup.

**Comment by byCedric** [comment]
Thanks for this! @vongohren

**Comment by nicolasrouanne** [comment]
@vongohren It works indeed 🎉 

Here is my complete configuration for reference:
```yaml
name: Expo Publish
on:
  push:
    branches:
      - master
jobs:
  publish:
    name: Install and publish
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v1
      - name: Setup kernel for react native, increase watchers
        run: echo fs.inotify.max_user_watches=524288 | sudo tee -a /etc/sysctl.conf && sudo sysctl -p
      - uses: actions/setup-node@v1
        with:
          node-version: 12.x
      - run: npm ci
      - uses: docker://bycedric/expo-cli-fix-inotify:latest
        with:
          args: publish
        env:
          EXPO_CLI_USERNAME: ${{ secrets.EXPO_CLI_USERNAME }}
          EXPO_CLI_PASSWORD: ${{ secrets.EXPO_CLI_PASSWORD }}
```

**Comment by nicolasrouanne** [comment]
@byCedric closing this, as there is a functional workaround

**Comment by hashim-sohail** [comment]
Getting the same error in my custom docker image that is being used in GitHub Actions. Can't modify system watchers directly as it is available in read-mode only. Does anyone have a fix?

**Comment by byCedric** [comment]
@hashim-sohail Haven't heard back from the GitHub dev team, so I'm afraid I have no solution. However, I bet if you post the inotify/watcher issue from Docker in the [community forum](https://github.community/t5/GitHub-Actions/bd-p/actions), you might have some more luck.

**Comment by byCedric** [comment]
I just released a new version with the suggested ENOSPC patch from Ville. Right now, you shouldn't encounter this issue with the config as described in the readme. You can see the patch being executed on Linux only, _and_ you can opt-out if you want 😄 Thank you all for helping ❤️

**Comment by kopax** [comment]
Hello, I am trying to run expo in docker gitlab-ci runner and I also have this error. I have tried to use in `before_script`:

```yaml
before_script:
  - echo fs.inotify.max_user_watches=524288 | tee -a /etc/sysctl.conf && sysctl -p
```

How did you guys finally fix this?

**Comment by byCedric** [comment]
@kopax I managed to integrate [the patch here](https://github.com/expo/expo-github-action/blob/master/src/system.ts#L19-L23). It's a bit GitHub-specific, you might have to fiddle your way around Gitlab though. If you can't get it working, hop on to the [Expo Developers slack](https://slack.expo.io/), maybe me (or someone else) can help you there 😄

**Comment by aadil42** [comment]
I'm getting ENOSPC error It says there's no space on device. What does it mean? Is it telling me that I have not enough space on my computer or does this mean something else? I'm using Linux

**Comment by aadil42** [comment]
I'm just trying to push to a dir on github.
